Add normalization to crosstab

nickeubank opened the pull request:
Closes #12569

Note does NOT address #12577

Commit:
add normalization to crosstab

sinhrks commented:
Thanks for the PR:)

As pointed in #12569, I prefer adding new ``aggfunc`` to new option (`normalize`). For example, isn't `agfunc="std"` and `normalize=True` meaningless?

nickeubank commented:
@sinhrks Yes, combining an `aggfunc` with normalize would be a problem, which is why I added an exception for that. 

(Or rather, I added two exceptions -- one for passing `aggfunc` without `values` which I think should raise an exception since when that happens `aggfunc` is ignored silently, and one for passing `normalize` with `values`. Together, they ensure no one tries to combine `normalize` with `values` or an `aggfunc`)

Since I can't think of an actual function you could pass to `aggfunc` that would do row or column normalization (unless I'm mistaken?), I'm not in favor of it being treated like it's just a special `aggfunc`. Since it's doing something you can't do with an `aggfunc`, seems like it should stand alone.

jreback commented:
@nickeubank, no you simply allow string args to ``aggfunc='normalize'``or whatever, its less confusing (and you can of course always pass an actual function)

jreback commented on pandas/tools/pivot.py:
```diff
@@ -434,12 +437,41 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+                table = table / table.sum(axis=1).sum(axis=0)
+
+            elif normalize == 'rows':
+                table = table.apply(lambda x: x / x.sum(), axis=1)
```
in any event this is just: ``table/x.sum(1)``. we almost *never* actually want to use ``.apply``

nickeubank commented:
@jreback @sinhrks ok, will update!

nickeubank commented:
@sinhrks little better?

jreback commented on pandas/tools/pivot.py:
```diff
@@ -381,8 +381,11 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
-    aggfunc : function, optional
+    aggfunc : function, or string, optional
         If no values array is passed, computes a frequency table
+        If 'pct' is passed, will normalize frequency table.
```
add a versionadded tag

sinhrks replied:
It may better to describe normalization is performed for the entire level (not for sub-levels) in ``MultiIndex`` case.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -434,12 +437,44 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
     data.update(zip(colnames, columns))
 
+    # Sanity checks
+    if values is None and callable(aggfunc):
```
can you show an example of this in master which fails?

nickeubank replied:
In [2]: df = pd.DataFrame({'a': [1, 2, 2, 2, 2], 'b': [3, 3, 4, 4, 4]})
    In [3]: pd.crosstab(df.a, df.b, aggfunc=np.mean) 
    Out[3]: 
    b  3  4
    a      
    1  1  0
    2  1  3

jreback replied:
so the problem is the user is passing something and it is ignored, ok then.

jreback replied:
make sure to note this in the whatsnew

nickeubank replied:
will do. Saving that till we decide on syntax.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -434,12 +437,44 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
                                aggfunc=len, margins=margins, dropna=dropna)
-        return table.fillna(0).astype(np.int64)
+
+        if isinstance(aggfunc, str):
```
this is very fragile. define functions and then have a dict lookup to do this.

nickeubank replied:
you mean instead of `if aggfunc == 'pct':`, define all those transformations into functions and then call `table = normalization_func(aggfunc)(table)` ?

jreback replied:
something like that

jreback replied:
even better would be to make a class out of ``crosstab``, then define these as functions on IT.

nickeubank replied:
ok, i'll give that some thought.

sinhrks commented on pandas/tools/pivot.py:
```diff
@@ -434,12 +437,44 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+        raise ValueError("Custom `aggfunc` functions require `values`"
+                         "to be specified")
+
+    if isinstance(aggfunc, str):
```
Pls use ``compat.string_types``

sinhrks commented:
I think it is nice if ``pct_xxx`` can normalize sum of values if ``values`` specified (and all of them are positive). 

```
a = np.array([1, 1, 2, 2])
b = np.array([1, 2, 1, 2])
c = np.array([10, 20, 40, 80])
pd.crosstab(a, b, c, aggfunc=np.sum)
# col_0   1   2
# row_0        
# 1      10  20
# 2      40  80

pd.crosstab(a, b, c, aggfunc='pct_cols')
# col_0   1    2
# row_0        
# 1     0.2  0.2
# 2     0.8  0.8
```

jorisvandenbossche commented:
I am slightly -1 on adding this functionality to `aggfunc`:

- First, having both the possibility of passing a func or a string (that not maps directly to an existing function) seems a bit unpythonic, and it is really about two different cases: how to aggregate multiple values, or whether or not to normalize the final result.
- Currently, you can already pass strings to `aggfunc` to specify the aggregation function (like in groupby.agg), so the current implementation checking for those specific strings will break that functionality.
- `value_counts` already has the `normalize` keyword, so there is a precedent for this (and I think it is an equivalent functionality, so would be more consistent to use that?)
- as @sinhrks just mentions, you may also want to normalize the result of an aggregation (eg `pd.crosstab(index, columns, values, aggfunc='sum', normalize=True`). That would not be possible with the current proposed API.

nickeubank commented:
Thanks @jorvisvandenbosse -- I think that's well said. Perhaps we should solicit a few more opinions to see if we can move towards consensus on this?

jreback commented:
@jorisvandenbossche soln seems reasonable.

nickeubank commented:
That would be my preference

jorisvandenbossche commented:
> you might want to enable normalize=True -> pct, normalize='index' -> pct_rows, normalize='columns' -> 'pct_cols'

+1

sinhrks commented:
My opinion is based on the understanding that ``crosstab`` is similar to ``pivot_table``. Thus:

- Because ``pivot_table`` accepts ``str``, I think it is OK to add other ``str`` options to ``aggfunc``.
- ``aggfunc="pct", "pct_cols", "pct_rows"`` outputs normalize counts if values is not specified, and normalize values if values is specified.
- I don't distinguish aggregate/normalize because ``aggfunc`` can be defined fornormalization. (``mean`` is ``(category sum)/(category counts)``, ``pct`` is ``(category sum)/(total counts)``)

I don't have strong opposition against @jorisvandenbossche 's option. One concern is API gets complex more than required. It's less likely to normalize other values than count and sum.

jorisvandenbossche commented:
> Because pivot_table accepts str, I think it is OK to add other str options to aggfunc

That's what I said above as well. `crosstab`'s `aggfunc` also accepts strings like `groupby().agg()` accepts strings. But there it are strings that match an existing function/method. This is not the case with 'pct_cols/rows'.

> One concern is API gets complex more than required.

Indeed, in general I am very reluctant in adding new keyword arguments. But in this case, I think it makes it more complex to explain what `aggfunc` does in case of combining it. It seems cleaner to separate the two notions

> It's less likely to normalize other values than count and sum.

That's indeed true.

nickeubank commented:
Sounds like we're agreed `aggfunc` can take strings (I'll remove that exception!)

I think my main two thoughts are:

- There is no function that could do row-normalization. Thus, since it's functionally doing something different than any explicit `aggfunc` could do, it should be it's own option, even though I agree with @sinhrks  there are few situations when you'd want to normalize the output of a crosstab generated with a custom `aggfunc`. Indeed, other programs like Stata that do something similar (accept aggregation functions for table) also split out normalization into its own syntax. 
- As @jorisvandenbossche has pointed out, `normalize` is already a keyword in a core function (`value_counts`), so I think it's pretty API consistent to have. 

Regarding complexity, I think of crosstab as a tool that may be used to generate analysis outputs to potentially put in papers (that's my interest at least). Given that, I think making it as flexible and powerful is highly desirable, even at the cost of an extra key word.

jreback commented:
so we are talking about doing these ops:

```
In [20]: r = pd.crosstab(df.a,df.b)

In [21]: r
Out[21]: 
b  3  4
a      
1  1  0
2  1  3

In [22]: r/r.sum()
Out[22]: 
b    3    4
a          
1  0.5  0.0
2  0.5  1.0

In [23]: r/r.sum().sum()
Out[23]: 
b    3    4
a          
1  0.2  0.0
2  0.2  0.6

In [24]: r.div(r.sum(1),axis='index')
Out[24]: 
b     3     4
a            
1  1.00  0.00
2  0.25  0.75
```

This kind of feels like a post-processing step rather than a result of a single operation.

jorisvandenbossche commented:
You can indeed rather simple do this after `crosstab` (although the row-wise case is not that trivial anymore), but still I would be OK to add a convenience keyword for this.

jreback commented:
yeah, just thinking if we have a kw (which is fine), then should explain what it is actually doing in the doc-string. As not completely obvious.

another option, is *maybe* to actually have a ``.normalize(...)`` function on frames. 

``pd.crosstab(df.a, df.b).normalize('columns')``

nickeubank commented:
Yes, it's post-processing. Indeed, that's how it's implemented. 

The one complication with a stand-alone normalize is that it can't be easily designed to deal with the `margin` columns -- if you do `crosstab(df.a, df.b, margins=True)` then you have to do a special normalization procedure to deal with the fact that the bottom row is a sum of all the rows above. A method that just divides all values by the total sum, you end up off by a factor of 1/2. I really thing there's value in having this as an integrated method.

jreback commented:
@nickeubank well one could argue the ``margins`` itself should be an external method as well! e.g. its kind of a 'footer' concept almost like a collection of DataFrame + footer Series + footer Series (rhs).

So that's really another abstraction and just being shoved into the current one.

nickeubank commented:
@jreback `margins` would only work as an external method for frequency tables -- for any other function (like mean), it's a set of `groupby` calls for different populations.

Also, say we made separate methods just for summing up values in each row or column called `add_margins` and had a separate `normalize`. It would still be impossible to make a normalized frequency table with margins. 

Say you had a frequency table `df` and called `df.add_margins().normalize()`, you'd end up with all the values off by a factor of 1/2. And if you called `df.normalize().add_margins()`, then unless you happened to have the same number of observations in each cell of `df`, the answers you'd get is wrong since you lose the "weight" of each cell coming from the number of observations when you normalize it.

jreback commented:
@nickeubank you missed my point. I was talking about a DIFFERENT abstraction, noticed I used the form: DataFrame + footer Series + footer Series (rhs), which in essence are additional data carried along (kind of like an Index). This doesn't exist ATM, and the use case is prob somewhat limited. My point is that it would work as the further methods would have access to the orginal data (and the margins), and then could easily normalize properly.

But that's just pie in the sky as it would require quite some effort.

nickeubank commented:
Ah, I see. Yes, that would be nice!

jorisvandenbossche commented:
In any case, regardless of this bit broader discussion, I think the original intent of this PR is OK? So @nickeubank I think you can certainlly update it.

nickeubank commented:
Sounds good. I'll update shortly!

nickeubank commented:
@jreback @jorisvandenbossche Updated. Comments on syntax? Will add a whatsnew when all else settled.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -393,6 +393,9 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
         Do not include columns whose entries are all NaN
+    normalize : boolean, or {'all','index','columns'}, default False
+        Provide percentages instead of frequencies. May not be
+        combined with values/aggfunc.
```
what does this last part mean?

nickeubank replied:
oops. We dropped that requirement, sorry.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +439,52 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+def _normalize(table, normalize, margins):
+
+    if isinstance(normalize, str) and \
+            normalize not in ['all', 'index', 'columns']:
```
use a dictionary where you define each of these operations as a lamba function.

Then if its not in the dict you can raise the invalid argument.

Then just actually execute it...something like:


```
normalizers = {
   'all' : lambda x: x / x.sum(axis=1).sum(axis=0),
.....
}
normalizers[True] = normalizers['all']

try:
   f = normalizes[normalize]
except KeyError:
    raise ValueError(.....)

table = f(table)
```

jreback commented on pandas/tools/pivot.py:
```diff
@@ -393,6 +393,12 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
         Do not include columns whose entries are all NaN
+    normalize : boolean, or {'all','index','columns'}, default False
+        Normalize by dividing all values by the sum of values.
+        If passed `all` or `True`, will normalize over all values.
```
put a blank line before the ``If passed`` and use ``-`` (e.g. a bulleted list) for each

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +443,56 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+    table = f(table)
+
+    if margins is True and normalize is not False:
+        table = table * 2
```
why do you do this?

nickeubank replied:
If the table has margins, then when you sum all the values you actually end up with 2x the sum of "real" cells. 

Ugh, though it now occurs to me that if we're gonna let people normalize the output of their own custom functions, that's not going to be robust. Only works for normalization of frequency tables. Not sure that can be fixed...

nickeubank replied:
Yeah, I don't think we can normalize custom functions if margin is used.

My inclination is to only allow `normalize` to be used if they don't pass a custom function. Thoughts?

Other option is to only allow them to `normalize` if either (a) no custom function, or (b) `margins=False`.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +443,56 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+        if normalize == "index":
+            table = table.drop('All', axis=1)
+        if normalize == "columns":
+            table = table.drop('All', axis=0)
```
you could do this in a little function table as well.

nickeubank replied:
Does that add much for only two cases?

jreback replied:
multiple ifs should either be if/else, or function tables. Function tables are more general and that's how we do it in the code-base.

jreback commented:
@nickeubank can divine what things are being done with custom aggfuncs? (e.g. see what tests do for this). Then you can see if it even makes sense to normalize for them.

nickeubank commented:
@jreback I'm not sure it's feasible to accommodate arbitrary functions passed to `aggfunc`. Among other things, because it's not always clear what normalization means outside of a frequency table. How do you normalize a mean? Or a standard deviation? Or the 90th percentile? 

Unless someone has a better ideal, I'm just going to restrict this to working when `aggfunc` is `None`.

nickeubank commented:
@sinhrks ok, now does frequency count (and will normalize counts) if passed `values` but no `aggfunc`, as you suggested!

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -18,6 +18,9 @@ Highlights include:
 ~~~~~~~~~~~~
 
+- ``pd.crosstab()`` now has ``normalize`` argument for normalizing frequency tables (:issue:`12569`).
+- ``pd.crosstab()`` returns count of non-missing observations in each group if passed ``values`` but no
```
this is a bug-fix (2nd one)

nickeubank replied:
not same as bug fix. They're different behaviors.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -393,6 +395,14 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+
+        - If passed `all` or `True`, will normalize over all values.
+        - If passed `index` will normalize over each row.
+        - If passed `columns` will normalize over each column.
```
directive versionadded

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,82 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
     data.update(zip(colnames, columns))
 
+    if values is None and aggfunc is not None:
+        raise ValueError("`aggfunc` cannot be used without `values`.")
```
don't need backticks

jreback commented on pandas/tools/pivot.py:
```diff
@@ -382,7 +382,9 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
         Values to group by in the columns
     values : array-like, optional
-        Array of values to aggregate according to the factors
+        Array of values to aggregate according to the factors.
```
this is the same thing as having ``aggfunc`` default to ``'count'``, so let's just make that explict if that's the case. Though I think it should just be an error like it was. Why are you changing this?

nickeubank replied:
As we discussed last week, `normalize` doesn't mix well with an `aggfunc`, but this is a clear situation where we'd want to be able to normalize values (as per @sinhrks suggestion). This makes that possible.

jreback replied:
but you are adding a default, that was not discussed. nor do I see a good reason for it. pls justify.

nickeubank replied:
Only kind of -- in the past, if `values` were specified but `aggfunc` was not, would throw an error. Now that situation defaults to `count`. But if you don't pass `values`, the old behavior persists. It does not change any established non-exception raising behaviors. 

The other option is to not allow users to normalize frequency tables of an arbitrary series of `values`. But seems like this is a pretty normal use case (and satisfies @sinhrks 's request that "I think it is nice if pct_xxx can normalize sum of values if values specified (and all of them are positive)."

jreback replied:
@nickeubank this is too magical the way you have done it. Pls remove this automatic ``aggfunc='count'``. you can propose this in another issue if you want.

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,82 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
-        return table.fillna(0).astype(np.int64)
+        table = table.fillna(0).astype(np.int64)
+
+    elif values is not None and aggfunc is None:
```
see my comment above

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,88 @@ def test_margin_ignore_dropna_bug(self):
         tm.assert_frame_equal(actual, expected)
 
+    def test_crosstab_normalize(self):
+        df = pd.DataFrame({'a': [1, 2, 2, 2, 2], 'b': [3, 3, 4, 4, 4],
```
add the issue number as a comment

jorisvandenbossche commented:
@nickeubank About the conflict between a custom aggfunc and normalize, I didn't follow the discussion of last week, but I don't see the problem. Shouldn't `normalize=True` not just ensure the values in the a row/column sum up to 1 (so are percentages of the sum), and I would say it's then up to the user to decide if that makes sense for his/her custom aggfunc?

nickeubank commented:
@jreback revised and rebased.

sinhrks commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,71 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+    normalizers = {
+        False: lambda x: x,
+        'all': lambda x: x / x.sum(axis=1).sum(axis=0),
+        'columns': lambda x: x / x.sum(),
```
It should accept 0/1 also to be compat with ``axis``?

nickeubank replied:
added

sinhrks commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,71 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+
+def _normalize(table, normalize, margins, aggfunc):
+
+    if aggfunc is not None and normalize is not False:
```
I think moving this check to ``crosstab`` and removing ``aggfunc`` from ``_normalize`` is clearer.

sinhrks commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,65 @@ def test_margin_ignore_dropna_bug(self):
+                                          margins=True), row_normal_margins)
+        tm.assert_frame_equal(pd.crosstab(df.a, df.b, normalize='columns',
+                                          margins=True), col_normal_margins)
+
```
Can you add test cases of 

- if ``index`` or ``columns`` are list of arrays/Series
- If sum of ``index`` or ``columns`` values includes 0 (current impl should be ``NaN``)

nickeubank replied:
@sinhrks  Could you please elaborate on those? Note sure I follow what you mean by first one. 

On second test, I don't think that can happen... If there are no observations in a row or column, that row or column doesn't appear in the table, does it?

sinhrks replied:
1st: Something like below.

```
pd.crosstab([np.array([1, 1, 2, 2]), np.array([1, 2, 1, 2])], np.array([1, 2, 1, 2]))
```

2nd: I assumed the case when user provides ``values``. 

```
pd.crosstab(np.array([1, 1, 2, 2]), np.array([1, 2, 1, 2]), np.array([0, 0, 0, 0]), aggfunc=np.sum)
```

jorisvandenbossche commented:
Quoting my own comment from above:

> @nickeubank About the conflict between a custom aggfunc and normalize, I didn't follow the discussion of last week, but I don't see the problem. Shouldn't normalize=True not just ensure the values in the a row/column sum up to 1 (so are percentages of the sum), and I would say it's then up to the user to decide if that makes sense for his/her custom aggfunc?

@nickeubank can you elaborate on this? 
It seems more a techinical issue due to how you implemented the `_normalize` function? (the `*2`) Wouldn't it be possible to alter the implementation by eg subsetting the frame to ignore the added margins when doing the normalization?

nickeubank commented:
@jorisvandenbossche Sorry missed that. Problems arise when `margins=True`.  

I thought about subsetting to main table, normalizing, and then separately normalizing the margins by summing and dividing by the sum. But that seems so strange, and would only actually mean anything for counts or sums. If someone passed `np.mean`, or `np.var`, it'd be generating all sorts of garbage. I suppose I can implement it, just seems strange to add a feature like that generates useless output in most settings. But I'm open to it if it would be preferred. 

I suppose I could also just say it can only be combined with `count` or  `sum`. The implementation where  if `aggfunc=None` and `values` was not `None` was meant as a compromise where the function would return a count of the passed values (analogous to how the default without `values` is a frequency table), but that was shot down. 

I'm open to suggestions, but most things I've tried implementing in this thread have made someone unhappy, so I was just going to settle for something conservative.

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -18,8 +18,7 @@ Highlights include:
 
-
-
+- ``pd.crosstab()`` now has ``normalize`` argument for normalizing frequency tables (:issue:`12569`).
```
I think you can add a sub-section here with an example / motivation (and add the same example to reshape.rst, with a pointer from here)

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,76 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
-        return table
+
+    # Post-process
+    table = _normalize(table, normalize=normalize, margins=margins)
```
```
if normalize:
   table = _normalize(......)
```

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,76 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+
+def _normalize(table, normalize, margins):
+
+    if not isinstance(normalize, bool) and not isinstance(normalize, str):
```
```
try:
     normalizer = normalizers[normalize]
except KeyError:
     raise ......
```

(and move below normalizeers)

jreback replied:
oh I see you have this below. So why do you have this section 476-481?

nickeubank replied:
476-481 were to deal with the 1s and 0s. I tried just putting them into the `normalizers` dict, but dicts treat `True` and `1` (and `False` and `0`) as the same when used as keys, so I had to deal with converting 1s and 0s before the dict. Otherwise if I try and do:

     normalizer[0] = normalizer['index']

I end up overwriting the entry for `normalizer[False]` and everything goes to haywire.

jreback replied:
what are ones and zeros for?

jreback replied:
oh, you mean like for the axis? 
try ``self._get_axis_number(axis)`` (self is a NDFrame object) which convert index/column to 0/1

nickeubank replied:
Don't think that solves the problem. Since the dict needs to have entires for "True" and "False", I can't use 0, 1 as axis arguments since they are treated the same.

sinhrks replied:
``isinstance(normalize, str)`` should use ``compcat.string_types``

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,76 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+
+    try:
+        f = normalizers[normalize]
+    except KeyError:
```
this should be enough

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,70 @@ def test_margin_ignore_dropna_bug(self):
+
+        df = pd.DataFrame({'a': [1, 2, 2, 2, 2], 'b': [3, 3, 4, 4, 4]})
+
+        with self.assertRaises(ValueError):
```
can you do assertRaisesRegex (as we have multiple ValueErrors we want to be sure the correct one is raised)

nickeubank commented:
@jorisvandenbossche : @sinhrks asked for something similar in comments to implementing as you suggested.

nickeubank commented:
@jorisvandenbossche @sinhrks Would you please check margin behavior (line 1017 in test_pivot.py) and let me know if that looks like what you would expect? Decided to internally normalize margins (so column margin normalized over values in column margin, index normalized over values in index margin), but could also do a normalization over sum of all margin values.

nickeubank commented:
@jreback getting an odd error in only one build (one of two python 2.7s). 

    AttributeError: 'TestCrosstab' object has no attribute 'assertRaisesRegex'

But runs fine in all the others. Is this on my side or a Travis config problem?

nickeubank commented:
@jreback Could you please advise -- should I remove the `assertRaisesRegex` tests or is this a travis issue?

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,124 @@ def test_margin_ignore_dropna_bug(self):
+                           'c': [1, 1, np.nan, 1, 1]})
+
+        error = 'values cannot be used without an aggfunc.'
+        with self.assertRaisesRegex(ValueError, error):
```
should be tm.assertRaisesRegex

sinhrks replied:
``tm.assertRaisesRegexp`` (ends with p)

nickeubank commented:
@sinhrks @jreback thanks! amended.

jreback commented on doc/source/reshaping.rst:
```diff
@@ -416,6 +421,45 @@ For example:
+    df
+
+    pd.crosstab(df.a, df.b)
+
```
versionadded tag here

jreback commented on doc/source/reshaping.rst:
```diff
@@ -402,6 +404,9 @@ It takes a number of arguments
   arrays passed
 - ``margins``: boolean, default False, Add row/column margins (subtotals)
+- ``normalize``: boolean, {"all","index","columns"}, or {0,1}, default False.
+   Normalize by dividing all values by the sum of values.
```
put spaces around args ``{'all', 'index', 'columns'}`` etc

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -1,4 +1,4 @@
-.. _whatsnew_0181:
+ .. _whatsnew_0181:
```
revert pls

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -18,8 +18,7 @@ Highlights include:
 
-
-
+- ``pd.crosstab()`` now has ``normalize`` argument for normalizing frequency tables (:issue:`12569`). Examples in updated docs :ref:`here <_reshaping.crosstabulations>`.
```
has gained. ref is ``<reshape.crosstabluations>`` (you don't use the ``_``)

jreback commented on pandas/tools/pivot.py:
```diff
@@ -393,6 +393,16 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
         Add row/column margins (subtotals)
     dropna : boolean, default True
         Do not include columns whose entries are all NaN
+    normalize : boolean, {'all','index','columns'}, or {0,1}, default False
```
spaces around ``all, index, columns``

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,98 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+        if normalize == 'columns':
+            column_margin = column_margin / column_margin.sum()
+            table = concat([table, column_margin], axis=1)
+            table.fillna(0)
```
this doesn't do anything, you need ``table = table.fillna(0)``. So pls add a test which catches this (I suspect you need an all nan column)

jreback commented on pandas/tools/pivot.py:
```diff
@@ -436,18 +446,98 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+            column_margin = column_margin / column_margin.sum()
+            index_margin = index_margin / index_margin.sum()
+            table = concat([table, column_margin], axis=1)
+            index_margin.loc['All'] = 1
```
move this 1 line up (better readability)

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,124 @@ def test_margin_ignore_dropna_bug(self):
 
+    def test_crosstab_normalize(self):
+        # Issue 12578
+        df = pd.DataFrame({'a': [1, 2, 2, 2, 2], 'b': [3, 3, 4, 4, 4],
```
addtl tests:

try these with all values for normalize & also do ``dropna=[True,False]`` (both values)
- all nan frame 
- empty frame

nickeubank commented:
@jreback Think addressed all your concerns. Only think missing is that I wasn't sure how to write some of the requested tests for `dropna=False` because I'm a little confused about expected behavior (see new comment on #10772).

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,159 @@ def test_margin_ignore_dropna_bug(self):
+                                            dtype='int64'),
+                             columns=pd.Index([3, 4], name='b'))
+        calculated = pd.crosstab(df.a, df.b, values=df.c, aggfunc='count',
+                                 normalize=True)
```
something this, just do in a list, as the tests are more clear

```
for arg in ['index', True, 'columns']:
     result = ....
     tm.assert_frame.....
```

jreback replied:
this too

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,159 @@ def test_margin_ignore_dropna_bug(self):
+                                 normalize=False)
+        tm.assert_frame_equal(nans, calculated)
+
+        df = pd.DataFrame({'a': [np.nan, np.nan, np.nan, np.nan, np.nan],
```
this last looks superfulous?

jreback commented on pandas/tools/tests/test_pivot.py:
```diff
@@ -965,6 +965,159 @@ def test_margin_ignore_dropna_bug(self):
+                                 normalize='columns')
+        tm.assert_frame_equal(empty, calculated)
+
+        nans = pd.DataFrame([[0.0, np.nan], [0.0, 0.0]],
```
do nans/empty in another function

jreback replied:
can you address this

jreback commented:
looks pretty good.

@sinhrks @jorisvandenbossche
cc @rosnfeld

jreback commented on doc/source/whatsnew/v0.18.1.txt:
```diff
@@ -18,8 +18,7 @@ Highlights include:
 
-
-
+- ``pd.crosstab()`` has gained ``normalize`` argument for normalizing frequency tables (:issue:`12569`). Examples in updated docs :ref:`here <reshaping.crosstabulations>`.
```
why don't you add a mini-example here (same one)

nickeubank replied:
@jreback. Thanks. I'm going to be traveling away from my computer for a
little over two weeks. I'm not sure what the timing is for 0.18.1. I'm
happy to make changes when I get back, but won't be able to do anything
till then.
On Mon, Apr 4, 2016 at 10:59 AM Jeff Reback <notifications@github.com>
wrote:

> In doc/source/whatsnew/v0.18.1.txt
> <https://github.com/pydata/pandas/pull/12578#discussion_r58418709>:
>
> > @@ -18,8 +18,7 @@ Highlights include:
> >  New features
> >  ~~~~~~~~~~~~
> >
> > -
> > -
> > +- ``pd.crosstab()`` has gained ``normalize`` argument for normalizing frequency tables (:issue:`12569`). Examples in updated docs :ref:`here <reshaping.crosstabulations>`.
>
> why don't you add a mini-example here (same one)
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/12578/files/39718b8f4f398d9cdf6c38583453c804675a5e6a#r58418709>
>

jreback commented:
@nickeubank when you are back, pls rebase. This looked pretty good.

nickeubank commented:
@jreback rebased!

nickeubank commented:
@jreback tweaked

jorisvandenbossche commented on doc/source/reshaping.rst:
```diff
@@ -402,6 +404,9 @@ It takes a number of arguments
   arrays passed
 - ``margins``: boolean, default False, Add row/column margins (subtotals)
+- ``normalize``: boolean, {'all', 'index', 'columns'}, or {0,1}, default False.
+   Normalize by dividing all values by the sum of values.
```
Very small detail, but I think there is one space too much at the beginning of this line

jorisvandenbossche commented on doc/source/reshaping.rst:
```diff
@@ -416,6 +421,47 @@ For example:
+.. versionadded:: 0.18.1
+
+Frequency tables can also be normalized to show percentages rather than counts 
+using the ``normalize`` command:
```
command -> argument ?

jorisvandenbossche commented on doc/source/reshaping.rst:
```diff
@@ -386,6 +386,8 @@ Note that ``pivot_table`` is also available as an instance method on DataFrame.
 Cross tabulations
 ~~~~~~~~~~~~~~~~~
 
+.. _reshaping.crosstabulations:
```
Can you put this before the title?

jorisvandenbossche commented on pandas/tools/pivot.py:
```diff
@@ -395,6 +395,16 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+    normalize : boolean, {'all', 'index', 'columns'}, or {0,1}, default False
+        Normalize by dividing all values by the sum of values.
+
+        - If passed `all` or `True`, will normalize over all values.
```
Can you put `'` around 'all'? Then the difference with True (string vs bool) is more clear I think

jorisvandenbossche replied:
and the same for items below

jorisvandenbossche commented on pandas/tools/pivot.py:
```diff
@@ -438,18 +448,97 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+        raise ValueError("aggfunc cannot be used without values.")
+
+    if values is not None and aggfunc is None:
+        raise ValueError("values cannot be used without an aggfunc.")
```
Can you maybe also add a notice of this in the docstring? For example in the explanationo of `values`: "``Need to specify `agfunc` ``"

nickeubank commented:
@jorisvandenbossche all integrated

sinhrks commented on pandas/tools/pivot.py:
```diff
@@ -438,18 +449,97 @@ def crosstab(index, columns, values=None, rownames=None, colnames=None,
+
+        # Actual Normalizations
+        normalizers = {
+            False: lambda x: x,
```
Never used, correct?

nickeubank commented:
@sinhrks ok! updated

jreback commented:
thanks @nickeubank
